[BUG] User Management edit user throws exception while saving email that is already in use.

<strong>Overview of the issue</strong>
User Management -> Edit User throws exception while saving email that is already in use. Same problem in Account -> Settings.

<pre>[ERROR] org.apache.catalina.core.ContainerBase.[Tomcat].[localhost].[/].[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [jhi_user_email_key]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement] with root cause
org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "jhi_user_email_key"
  Detail: Key (email)=(user@localhost) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2182) ~[postgresql-9.4-1203-jdbc42.jar:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1911) ~[postgresql-9.4-1203-jdbc42.jar:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:173) ~[postgresql-9.4-1203-jdbc42.jar:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:618) ~[postgresql-9.4-1203-jdbc42.jar:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:468) ~[postgresql-9.4-1203-jdbc42.jar:9.4]</pre>

<strong>JHipster Version(s)</strong>
Latest: 2.23.0

<strong>Reproduce the error</strong>
Try to save email that is used by another user. For example while logged in as admin, try to update your email to user@localhost in the sample app.

<strong>Suggest a Fix</strong>
Add same email validations to update user and update account endpoints that are used in register account endpoint.
<pre>.map(user -> new ResponseEntity<>("e-mail address already in use", HttpStatus.BAD_REQUEST))</pre>

--- a/app/templates/src/main/java/package/web/rest/_AccountResource.java
+++ b/app/templates/src/main/java/package/web/rest/_AccountResource.java
@@ -123,17 +123,16 @@ public class AccountResource {
         produces = MediaType.APPLICATION_JSON_VALUE)
     @Timed
     public ResponseEntity<String> saveAccount(@RequestBody UserDTO userDTO) {
+        Optional<User> existingUser = userRepository.findOneByEmail(userDTO.getEmail());
+        if (existingUser.isPresent() && (!existingUser.get().getLogin().equalsIgnoreCase(userDTO.getLogin()))) {
+            return ResponseEntity.badRequest().header("Failure", "Email already used").body(null);
+        }
         return userRepository
             .findOneByLogin(SecurityUtils.getCurrentUser().getUsername())
             .map(u -> {
-                Optional<User> oUser=userRepository.findOneByEmail(userDTO.getEmail());
-                if ((!oUser.isPresent()) || (oUser.isPresent() && oUser.get().getLogin().equalsIgnoreCase(u.getLogin()))) {
-                  userService.updateUserInformation(userDTO.getFirstName(), userDTO.getLastName(), userDTO.getEmail(),
-                      userDTO.getLangKey());
-                  return new ResponseEntity<String>(HttpStatus.OK);
-                }else{
-                  return new ResponseEntity<String>(HttpStatus.INTERNAL_SERVER_ERROR);
-                }
+                userService.updateUserInformation(userDTO.getFirstName(), userDTO.getLastName(), userDTO.getEmail(),
+                    userDTO.getLangKey());
+                return new ResponseEntity<String>(HttpStatus.OK);
             })
             .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
     }

--- a/app/templates/src/main/java/package/web/rest/_AccountResource.java
+++ b/app/templates/src/main/java/package/web/rest/_AccountResource.java
@@ -126,9 +126,14 @@ public class AccountResource {
         return userRepository
             .findOneByLogin(SecurityUtils.getCurrentUser().getUsername())
             .map(u -> {
-                userService.updateUserInformation(userDTO.getFirstName(), userDTO.getLastName(), userDTO.getEmail(),
-                    userDTO.getLangKey());
-                return new ResponseEntity<String>(HttpStatus.OK);
+                Optional<User> oUser=userRepository.findOneByEmail(userDTO.getEmail());
+                if ((!oUser.isPresent()) || (oUser.isPresent() && oUser.get().getLogin().equalsIgnoreCase(u.getLogin()))) {
+                  userService.updateUserInformation(userDTO.getFirstName(), userDTO.getLastName(), userDTO.getEmail(),
+                      userDTO.getLangKey());
+                  return new ResponseEntity<String>(HttpStatus.OK);
+                }else{
+                  return new ResponseEntity<String>(HttpStatus.INTERNAL_SERVER_ERROR);
+                }
             })
             .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
     }

--- a/app/templates/src/main/java/package/web/rest/_UserResource.java
+++ b/app/templates/src/main/java/package/web/rest/_UserResource.java
@@ -105,38 +105,37 @@ public class UserResource {
     @Secured(AuthoritiesConstants.ADMIN)
     public ResponseEntity<ManagedUserDTO> updateUser(@RequestBody ManagedUserDTO managedUserDTO) throws URISyntaxException {
         log.debug("REST request to update User : {}", managedUserDTO);
-        Optional<User> oUser=userRepository.findOneByEmail(managedUserDTO.getEmail());
-        if ((!oUser.isPresent()) || (oUser.isPresent() && oUser.get().getLogin().equalsIgnoreCase(managedUserDTO.getLogin()))) {
-          return userRepository
-              .findOneById(managedUserDTO.getId())
-              .map(user -> {
-                  user.setLogin(managedUserDTO.getLogin());
-                  user.setFirstName(managedUserDTO.getFirstName());
-                  user.setLastName(managedUserDTO.getLastName());
-                  user.setEmail(managedUserDTO.getEmail());
-                  user.setActivated(managedUserDTO.isActivated());
-                  user.setLangKey(managedUserDTO.getLangKey());<% if (databaseType == 'sql' || databaseType == 'mongodb') { %>
-                  Set<Authority> authorities = user.getAuthorities();
-                  authorities.clear();
-                  managedUserDTO.getAuthorities().stream().forEach(
-                      authority -> authorities.add(authorityRepository.findOne(authority))
-                  );<% if (databaseType == 'mongodb') { %>
-                  userRepository.save(user);<% } %>
-                  return ResponseEntity.ok()
-                      .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
-                      .body(new ManagedUserDTO(userRepository
-                          .findOne(managedUserDTO.getId())));<% } else { %>
-                  user.setAuthorities(managedUserDTO.getAuthorities());
-                  userRepository.save(user);
-                  return ResponseEntity.ok()
-                      .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
-                      .body(new ManagedUserDTO(userRepository
-                          .findOne(managedUserDTO.getId())));<% } %>
-              })
-              .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
-        }else{
-          return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        Optional<User> existingUser = userRepository.findOneByEmail(managedUserDTO.getEmail());
+        if (existingUser.isPresent() && (!existingUser.get().getLogin().equalsIgnoreCase(managedUserDTO.getLogin()))) {
+            return ResponseEntity.badRequest().header("Failure", "Email already used").body(null);
         }
+        return userRepository
+            .findOneById(managedUserDTO.getId())
+            .map(user -> {
+                user.setLogin(managedUserDTO.getLogin());
+                user.setFirstName(managedUserDTO.getFirstName());
+                user.setLastName(managedUserDTO.getLastName());
+                user.setEmail(managedUserDTO.getEmail());
+                user.setActivated(managedUserDTO.isActivated());
+                user.setLangKey(managedUserDTO.getLangKey());<% if (databaseType == 'sql' || databaseType == 'mongodb') { %>
+                Set<Authority> authorities = user.getAuthorities();
+                authorities.clear();
+                managedUserDTO.getAuthorities().stream().forEach(
+                    authority -> authorities.add(authorityRepository.findOne(authority))
+                );<% if (databaseType == 'mongodb') { %>
+                userRepository.save(user);<% } %>
+                return ResponseEntity.ok()
+                    .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
+                    .body(new ManagedUserDTO(userRepository
+                        .findOne(managedUserDTO.getId())));<% } else { %>
+                user.setAuthorities(managedUserDTO.getAuthorities());
+                userRepository.save(user);
+                return ResponseEntity.ok()
+                    .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
+                    .body(new ManagedUserDTO(userRepository
+                        .findOne(managedUserDTO.getId())));<% } %>
+            })
+            .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
 
     }
 

--- a/app/templates/src/main/java/package/web/rest/_UserResource.java
+++ b/app/templates/src/main/java/package/web/rest/_UserResource.java
@@ -105,33 +105,39 @@ public class UserResource {
     @Secured(AuthoritiesConstants.ADMIN)
     public ResponseEntity<ManagedUserDTO> updateUser(@RequestBody ManagedUserDTO managedUserDTO) throws URISyntaxException {
         log.debug("REST request to update User : {}", managedUserDTO);
-        return userRepository
-            .findOneById(managedUserDTO.getId())
-            .map(user -> {
-                user.setLogin(managedUserDTO.getLogin());
-                user.setFirstName(managedUserDTO.getFirstName());
-                user.setLastName(managedUserDTO.getLastName());
-                user.setEmail(managedUserDTO.getEmail());
-                user.setActivated(managedUserDTO.isActivated());
-                user.setLangKey(managedUserDTO.getLangKey());<% if (databaseType == 'sql' || databaseType == 'mongodb') { %>
-                Set<Authority> authorities = user.getAuthorities();
-                authorities.clear();
-                managedUserDTO.getAuthorities().stream().forEach(
-                    authority -> authorities.add(authorityRepository.findOne(authority))
-                );<% if (databaseType == 'mongodb') { %>
-                userRepository.save(user);<% } %>
-                return ResponseEntity.ok()
-                    .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
-                    .body(new ManagedUserDTO(userRepository
-                        .findOne(managedUserDTO.getId())));<% } else { %>
-                user.setAuthorities(managedUserDTO.getAuthorities());
-                userRepository.save(user);
-                return ResponseEntity.ok()
-                    .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
-                    .body(new ManagedUserDTO(userRepository
-                        .findOne(managedUserDTO.getId())));<% } %>
-            })
-            .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
+        Optional<User> oUser=userRepository.findOneByEmail(managedUserDTO.getEmail());
+        if ((!oUser.isPresent()) || (oUser.isPresent() && oUser.get().getLogin().equalsIgnoreCase(managedUserDTO.getLogin()))) {
+          return userRepository
+              .findOneById(managedUserDTO.getId())
+              .map(user -> {
+                  user.setLogin(managedUserDTO.getLogin());
+                  user.setFirstName(managedUserDTO.getFirstName());
+                  user.setLastName(managedUserDTO.getLastName());
+                  user.setEmail(managedUserDTO.getEmail());
+                  user.setActivated(managedUserDTO.isActivated());
+                  user.setLangKey(managedUserDTO.getLangKey());<% if (databaseType == 'sql' || databaseType == 'mongodb') { %>
+                  Set<Authority> authorities = user.getAuthorities();
+                  authorities.clear();
+                  managedUserDTO.getAuthorities().stream().forEach(
+                      authority -> authorities.add(authorityRepository.findOne(authority))
+                  );<% if (databaseType == 'mongodb') { %>
+                  userRepository.save(user);<% } %>
+                  return ResponseEntity.ok()
+                      .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
+                      .body(new ManagedUserDTO(userRepository
+                          .findOne(managedUserDTO.getId())));<% } else { %>
+                  user.setAuthorities(managedUserDTO.getAuthorities());
+                  userRepository.save(user);
+                  return ResponseEntity.ok()
+                      .headers(HeaderUtil.createEntityUpdateAlert("user", managedUserDTO.getLogin()))
+                      .body(new ManagedUserDTO(userRepository
+                          .findOne(managedUserDTO.getId())));<% } %>
+              })
+              .orElseGet(() -> new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR));
+        }else{
+          return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
     }
 
     /**

